### ISSUE TITLE
ActiveRecord: Wrong results returned by ActiveRecord::Associations::Preloader when intermediate association is already preloaded.

### ISSUE THREAD

**Issue by yopzer** [rails/rails#45822]
### Steps to reproduce
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "activerecord", "7.0.3.1"
  gem "sqlite3"
end

require "active_record"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :a_records, force: true

  create_table :b_records, force: true do |t|
    t.references :a_record
    t.references :c_record
  end

  create_table :c_records, force: true

  create_table :d_records, force: true do |t|
    t.references :c_record
  end
end

class ARecord < ActiveRecord::Base
  has_one :b_record
  has_one :c_record, through: :b_record
  has_many :d_records, through: :c_record
end

class BRecord < ActiveRecord::Base
  belongs_to :a_record
  belongs_to :c_record
end

class CRecord < ActiveRecord::Base
  has_one :b_record
  has_many :d_records
end

class DRecord < ActiveRecord::Base
  belongs_to :c_record
end

a = ARecord.create!
c = CRecord.create!
BRecord.create!(a_record: a, c_record: c)
DRecord.create!(c_record: c)
DRecord.create!(c_record: c)

a.reload
a.c_record # NB: if removed this works

::ActiveRecord::Associations::Preloader.new(records: [a], associations: :d_records).call
puts({ record_ids: a.d_records.map(&:id), size: a.d_records.size })
```

### Expected behavior
Two records should be loaded:
`{:record_ids=>[1, 2], :size=>2}`

### Actual behavior
No records are loaded:
`{:record_ids=>[], :size=>0}`

### System configuration
**Rails version**: 7.0.3.1

**Ruby version**: 2.7.5

**Comment by yopzer** [comment]
Identified the change after it started behaving differently: https://github.com/rails/rails/pull/43137

**Comment by MatElGran** [comment]
The test still fails on main

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :a_records, force: true

  create_table :b_records, force: true do |t|
    t.references :a_record
    t.references :c_record
  end

  create_table :c_records, force: true

  create_table :d_records, force: true do |t|
    t.references :c_record
  end
end

class ARecord < ActiveRecord::Base
  has_one :b_record
  has_one :c_record, through: :b_record
  has_many :d_records, through: :c_record
end

class BRecord < ActiveRecord::Base
  belongs_to :a_record
  belongs_to :c_record
end

class CRecord < ActiveRecord::Base
  has_one :b_record
  has_many :d_records
end

class DRecord < ActiveRecord::Base
  belongs_to :c_record
end

class BugTest < Minitest::Test
  def test_association_stuff
    a = ARecord.create!
    c = CRecord.create!
    BRecord.create!(a_record: a, c_record: c)
    DRecord.create!(c_record: c)
    DRecord.create!(c_record: c)

    a.reload
    a.c_record # NB: if removed this works

    ::ActiveRecord::Associations::Preloader.new(records: [a], associations: :d_records).call
    assert_equal 2, a.d_records.size
  end
end
```

**Comment by joshuay03** [comment]
Hey all,

I decided to investigate this and found that root cause is in `ActiveRecord::Associations::Preloader::ThroughAssociation#records_by_owner` where different instances of through_records were being used to fetch the memoized source_records_by_owner. I've got a PR up here with a failing test and a band aid fix which currently makes it pass: https://github.com/rails/rails/pull/46579.

I'm unsure what the best fix is though since I've never dived into this area before. Any thoughts?
